File: Claude/Python/2024/Day2b_Solution_2024.py
def is_safe(levels):
    diffs = [levels[i+1] - levels[i] for i in range(len(levels)-1)]
    return all(1 <= d <= 3 for d in diffs) or all(-3 <= d <= -1 for d in diffs)

def is_safe_dampened(levels):
    if is_safe(levels):
        return True
    return any(is_safe(levels[:i] + levels[i+1:]) for i in range(len(levels)))

with open("Day2_Input.txt") as f:
    reports = [list(map(int, line.split())) for line in f]

print(sum(is_safe_dampened(r) for r in reports))

File: Claude/Rust/2024/Day2b_Solution_2024.rs
use std::fs;

fn is_safe(levels: &[i64]) -> bool {
    if levels.len() < 2 {
        return true;
    }
    let increasing = levels[1] > levels[0];
    levels.windows(2).all(|w| {
        let diff = w[1] - w[0];
        if increasing { diff >= 1 && diff <= 3 } else { diff <= -1 && diff >= -3 }
    })
}

fn is_safe_with_dampener(levels: &[i64]) -> bool {
    if is_safe(levels) {
        return true;
    }
    (0..levels.len()).any(|i| {
        let reduced: Vec<i64> = levels.iter().enumerate().filter(|&(j, _)| j != i).map(|(_,&v)| v).collect();
        is_safe(&reduced)
    })
}

fn main() {
    let input = fs::read_to_string("Day2_Input.txt").expect("Failed to read input file");

    let count = input.lines()
        .filter(|line| !line.is_empty())
        .filter(|line| {
            let levels: Vec<i64> = line.split_whitespace().map(|n| n.parse().unwrap()).collect();
            is_safe_with_dampener(&levels)
        })
        .count();

    println!("{}", count);
}
